### DISCUSSION TITLE
Kubernetes 1.29 Major Themes

### DISCUSSION THREAD

**Discussion by fsmunoz** [kubernetes/sig-release#2349] [General]
This is a discussion to discuss Major Themes for the v1.29 release.

Here is the [discussion thread](https://github.com/kubernetes/sig-release/discussions/2271) from the last release cycle.

cc @kubernetes/release-team-leads 
cc @fykaa @mengjiao-liu @muddapu @rashansmith

**Comment by sanposhiho** [comment]
From SIG-Scheduling:

- https://github.com/kubernetes/enhancements/issues/4247
  - The throughput of the scheduler is our eternal challenge. This QueueingHint feature brings a new possibility to optimize the efficiency of requeueing, which could reduce useless scheduling retries significantly. We're actively working on this, and some stable in-tree plugins will have supported it in v1.29.
- https://github.com/kubernetes/enhancements/issues/3633
  - One enhancement will be introduced in PodAffinity/PodAntiAffinity as alpha. It will increase the accuracy of calculation during rolling updates.
- https://github.com/kubernetes/enhancements/issues/2485
  - [ReadWriteOncePod access mode](https://kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha/) for PV is moving towards GA.
- https://github.com/kubernetes/enhancements/issues/3902
  - It's marked as beta, but actually a completely new change. (It skips the alpha.)
  - As title describes, it's to decouple `TaintManager` that performs taint-based pod eviction from `NodeLifecycleController` and make them two separate controllers: `NodeLifecycleController` to add taints to unhealthy nodes and `TaintManager` to perform pod deletion on nodes tainted with NoExecute effect.

@kubernetes/sig-scheduling-approvers Please follow up if anything missing.

**Reply by alculquicondor** [reply]
That's a great summary, thank you @sanposhiho

**Reply by justankiit** [reply]
That is a great Summary @sanposhiho

**Reply by kerthcet** [reply]
Thanks @sanposhiho for the effort.

**Comment by xing-yang** [comment]
From SIG-Storage:
- [ReadWriteOncePod Persistent Volume Access Mode](https://github.com/kubernetes/enhancements/issues/2485) feature moved to GA in v1.29. This feature introduces a new ReadWriteOncePod access mode for PersistentVolumes that restricts access to a single pod on a single node at a time.
- [NodeExpandSecret](https://github.com/kubernetes/enhancements/issues/3107) feature moved to GA in v1.29. This feature adds NodeExpandSecret to the CSI persistent volume source and enables the CSI client to send it out as part of the nodeExpandVolume request to the CSI drivers.
- The [VolumeAttributesClass](https://github.com/kubernetes/enhancements/issues/3751) Alpha feature was introduced in v1.29. This feature proposes an extension to the Kubernetes Persistent Volume API to allow users to change volume attributes after provisioning (for example, for IOPS or throughput).

cc @saad-ali @msau42 @jsafrane

**Reply by carlory** [reply]
In 1.29, the VolumeAttributesClass feature maybe only contain the API changes and its functions are not implemented yet. So, I'm not sure whether we should add an announcement in the major themes. 

cc @msau42 @sunnylovestiramisu

**Reply by xing-yang** [reply]
Implementation is in external-provisioner and external-resizer which will be released async after K8s 1.29 release.

**Comment by aojea** [comment]
From SIG-Network:

* [nftables kube-proxy backend](https://github.com/kubernetes/enhancements/issues/3866) Alpha feature introduced in v1.29. This features adds a new backend to kube-proxy based on nftables, since kernel community is focused mainly in nftables and practically stopped the iptables development
* [Multiple Service CIDRs](https://github.com/kubernetes/enhancements/issues/1880). Alpha feature introduced in v1.29. This features allow cluster administrators to dynamically resize the service IP ranges assigned to their clusters to deal with problems like IP exhaustion or IP renumbering.

cc: @danwinship @thockin @shaneutt

**Comment by aravindhp** [comment]
From SIG-Windows:
- https://github.com/kubernetes/enhancements/issues/4216
  - Alpha feature introduced in v1.29
- https://github.com/kubernetes/enhancements/issues/1287
  - Windows support going Alpha in v1.29

**Comment by elmiko** [comment]
From SIG-Cloud Provider:
* https://github.com/kubernetes/enhancements/pull/4190
  * feature moved to beta status, users need to be aware that the feature gates `DisableCloudProviders` and `DisableKubeletCloudCredentialProvider` will now default to true. This means that users on Azure, GCE, and vSphere who are not using external cloud controller managers will need to set these feature gates to false, or migrate to external CCMs.

:warning:  **This has the potential to break users who are not prepared for the change in feature gates.**

cc @bridgetkromhout @andrewsykim @cheftako

**Comment by logicalhan** [comment]
From SIG-Instrumentation:

- https://github.com/kubernetes/kubernetes/pull/116897
  - [Kubelet resource metrics](https://kubernetes.io/docs/reference/instrumentation/node-metrics/) are graduated to GA
- https://github.com/kubernetes/kubernetes/pull/120574
  - [Component SLI metrics](https://kubernetes.io/docs/reference/instrumentation/slis/) are graduated to GA
- https://github.com/kubernetes/enhancements/issues/3077
  - kube-scheduler was converted to contextual logging

**Comment by mrunalp** [comment]
From SIG-Node: 
* https://github.com/kubernetes/kubernetes/pull/121579 Sidecar containers beta
* https://github.com/kubernetes/enhancements/issues/3673 Max image parallel pulls beta
* https://github.com/kubernetes/kubernetes/pull/118760 User namespaces pod security alpha
* https://github.com/kubernetes/enhancements/issues/4216 Image pull per runtime class alpha
* https://github.com/kubernetes/enhancements/issues/4191 Split filesystem alpha
* https://github.com/kubernetes/enhancements/issues/4210 Kubelet image GC after maximum age alpha

**Reply by SergeyKanzhelev** [reply]
For sidecars it is important to mention that if sidecars feature gate is enabled and used with API server 1.29+ while node is 1.27 and below - Pod will not work. So the feature should only be used when all nodes were upgraded to 1.28+

**Reply by SergeyKanzhelev** [reply]
We also want to highlight that Evente PLEG beta feature that is disabled by default have a critical bug: https://github.com/kubernetes/kubernetes/issues/121349#issuecomment-1813029991 that is not resolved yet. We recommend to use extra caution when enabling this feature.

**Reply by liggitt** [reply]
> We also want to highlight that Evente PLEG beta feature that is disabled by default have a critical bug: https://github.com/kubernetes/kubernetes/issues/121349#issuecomment-1813029991 that is not resolved yet. We recommend to use extra caution when enabling this feature.

That's a good candidate for the "Known Issues" section of the release notes

**Reply by SergeyKanzhelev** [reply]
Is there a mechanism to include it there?

**Reply by liggitt** [reply]
add to https://github.com/kubernetes/kubernetes/issues/121886

**Reply by pacoxu** [reply]
> https://github.com/kubernetes/enhancements/issues/3673

I added a new e2e in https://github.com/kubernetes/kubernetes/pull/121604 which is not merged yet. The adding e2e is a beta criterion for this feature. However, there is no FG for this KEP.

**Comment by ritazh** [comment]
From SIG-Auth:
- https://github.com/kubernetes/kubernetes/pull/121485 
  - [3299](https://kep.k8s.io/3299) KMS v2 encryption at rest improvement graduates to GA in 1.29
- https://github.com/kubernetes/kubernetes/pull/120682
  - [2799](https://kep.k8s.io/2799) Reduce secret based SA token, LegacyServiceAccountTokenCleanUp going to Beta in 1.29
- https://github.com/kubernetes/kubernetes/issues/118872
  - [3221](https://kep.k8s.io/3221) Structured Authorization Configuration going Alpha in 1.29
- https://github.com/kubernetes/kubernetes/pull/119142
  - [3331](https://kep.k8s.io/3331) Structured OIDC Configuration going Alpha in 1.29
- https://github.com/kubernetes/kubernetes/pull/113374
  - [3257](https://kep.k8s.io/3257) Cluster Trust Bundles going Alpha in 1.29
- https://github.com/kubernetes/kubernetes/pull/120780
  - [4193](https://kep.k8s.io/4193) Bound service account token improvements going Alpha in 1.29


cc @liggitt @enj @aramase @ahmedtd @munnerz @nilekhc @palnabarun @yt2985

**Reply by liggitt** [reply]
If we wanted a focused list, the KMS GA and token cleanup beta are the ones I'd highlight. The alpha stuff is interesting as a preview of what's coming but is admin-focused or not enabled by default.

**Comment by ameukam** [comment]
From SIG Release:
 - apt.kubernetes.io & yum.kubernetes.io are going away. See https://kubernetes.io/blog/2023/08/31/legacy-package-repository-deprecation/

**Reply by BenTheElder** [reply]
Can we also clarify the backing kubernetes packages on `packages.cloud.google.com` as mentioned in https://github.com/kubernetes/website/pull/44109#issue-2012931936

"apt.kubernetes.io & yum.kubernetes.io AKA packages.cloud.google.com legacy Kubernetes packages are going away, to be replaced by pkgs.k8s.io. See: https://kubernetes.io/blog/2023/08/31/legacy-package-repository-deprecation/"

**Reply by mrbobbytables** [reply]
+1 from me

**Comment by krol3** [comment]
Hi Team, here Carol from the Communication 1.29, we will like to choose the top 3 major themes for the webinar 1.29. What do you think? @kubernetes/sig-release-leads cc: @Priyankasaggu11929 @kubernetes/release-team-comms

- [KEP-2395](https://github.com/kubernetes/enhancements/pull/4190) Update 2395-removing-in-tree-cloud-providers to beta status
- [KEP-3299](https://github.com/kubernetes/enhancements/issues/3299) KMS v2 Improvements #3299
- [KEP-1287](https://github.com/kubernetes/enhancements/issues/1287) In-Place Update of Pod Resources #1287

**Reply by elmiko** [reply]
+1 for kep-2395 being in there
